bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java

```diff
@@ -0,0 +1,205 @@
+                ItemData[] input = new ItemData[size];
+                for (int i = 0; i < input.length; i++) {
+                    // Index is converting char to integer, adding i then converting back to char based on ASCII code
+                    Pair<ItemStack, ItemData> result = translateToBedrock(session, shapedRecipe.getIngredientMap().get((char) ('a' + i)));
+                    ingredients[i] = new Ingredient(new ItemStack[]{result.getKey()});
+                    input[i] = result.getValue();
+                    isNotAllAir |= input[i].getId() != 0;
+                }
+
+                if (!isNotAllAir) continue;
+                UUID uuid = UUID.randomUUID();
+                // Add recipe to our internal cache
+                ShapedRecipeData data = new ShapedRecipeData(shapedRecipe.getShape()[0].length(), shapedRecipe.getShape().length,
+                        "", ingredients, javaOutput);
+                session.getCraftingRecipes().put(netId,
+                        new com.github.steveice10.mc.protocol.data.game.recipe.Recipe(RecipeType.CRAFTING_SHAPED, uuid.toString(), data));
+
+                // Add recipe for Bedrock
+                craftingDataPacket.getCraftingData().add(CraftingData.fromShaped(uuid.toString(),
+                        shapedRecipe.getShape()[0].length(), shapedRecipe.getShape().length, Arrays.asList(input),
+                        Collections.singletonList(output), uuid, "crafting_table", 0, netId++));
+            } else if (recipe instanceof ShapelessRecipe) {
+                ShapelessRecipe shapelessRecipe = (ShapelessRecipe) recipe;
+                Ingredient[] ingredients = new Ingredient[shapelessRecipe.getIngredientList().size()];
+                ItemData[] input = new ItemData[shapelessRecipe.getIngredientList().size()];
+
+                for (int i = 0; i < input.length; i++) {
+                    Pair<ItemStack, ItemData> result = translateToBedrock(session, shapelessRecipe.getIngredientList().get(i));
+                    ingredients[i] = new Ingredient(new ItemStack[]{result.getKey()});
+                    input[i] = result.getValue();
+                    isNotAllAir |= input[i].getId() != 0;
+                }
+
+                if (!isNotAllAir) continue;
+                UUID uuid = UUID.randomUUID();
+                // Add recipe to our internal cache
+                ShapelessRecipeData data = new ShapelessRecipeData("", ingredients, javaOutput);
+                session.getCraftingRecipes().put(netId,
+                        new com.github.steveice10.mc.protocol.data.game.recipe.Recipe(RecipeType.CRAFTING_SHAPELESS, uuid.toString(), data));
+
+                // Add recipe for Bedrock
+                craftingDataPacket.getCraftingData().add(CraftingData.fromShapeless(uuid.toString(),
+                        Arrays.asList(input), Collections.singletonList(output), uuid, "crafting_table", 0, netId++));
+            }
+        }
+
```

Comment:
Could this be stored as a constant? Unless there is something I'm missing, it seems very expensive to rerun this code every time a player joins.